components/provider/src/lib.rs

```diff
@@ -89,7 +89,155 @@ pub mod prelude {
+        borrowed.map(|any| {
+            let actual_type_id = (any as &dyn Any).type_id();
+            any.downcast_mut().ok_or_else(|| Error::MismatchedType {
+                actual: actual_type_id,
+                generic: Some(TypeId::of::<T>()),
+            })
+        })
+    }
+
+    #[derive(Debug)]
+    pub struct DataReceiverImpl<'d, T>
+    where
+        T: serde::Deserialize<'static> + erased_serde::Serialize + Any + Clone + Debug,
+    {
+        pub payload: Option<Cow<'d, T>>,
+    }
+
+    impl<'d, T> DataReceiver<'d, 'static> for DataReceiverImpl<'d, T>
+    where
+        T: serde::Deserialize<'static> + erased_serde::Serialize + Any + Clone + Debug,
+    {
+        fn set_to(
+            &mut self,
+            deserializer: &mut dyn erased_serde::Deserializer<'static>,
+        ) -> Result<(), erased_serde::Error> {
+            let obj: T = erased_serde::deserialize(deserializer)?;
+            self.payload = Some(Cow::Owned(obj));
+            Ok(())
+        }
+
+        fn set_to_any(&mut self, any: &'d dyn Any) -> Result<(), Error> {
+            self.payload = Some(Cow::Borrowed(any.downcast_ref().ok_or_else(|| {
+                Error::MismatchedType {
+                    actual: any.type_id(),
+                    generic: Some(TypeId::of::<T>()),
+                }
+            })?));
+            Ok(())
+        }
+
+        fn set_to_boxed(&mut self, boxed: Box<dyn Any>) -> Result<(), Error> {
+            self.payload = Some(Cow::Owned(
+                *(boxed.downcast().map_err(|any| Error::MismatchedType {
+                    actual: any.type_id(),
+                    generic: Some(TypeId::of::<T>()),
+                })?),
+            ));
+            Ok(())
+        }
+
+        fn borrow_payload_as_any(&self) -> Option<&dyn Any> {
+            match &self.payload {
+                Some(cow) => {
+                    let borrowed: &T = cow.borrow();
+                    Some(borrowed.as_any())
+                }
+                None => None,
+            }
+        }
+
+        fn borrow_payload_as_serialize(&self) -> Option<&dyn erased_serde::Serialize> {
+            match &self.payload {
+                Some(cow) => {
+                    let borrowed: &T = cow.borrow();
+                    Some(borrowed as &dyn erased_serde::Serialize)
+                }
+                None => None,
+            }
+        }
+
+        fn borrow_payload_as_any_mut(&mut self) -> Option<&mut dyn Any> {
+            match &mut self.payload {
+                Some(cow) => {
+                    let borrowed: &mut T = cow.to_mut().borrow_mut();
+                    Some(borrowed.as_any_mut())
+                }
+                None => None,
+            }
+        }
+    }
+
+    #[derive(Debug, Clone)]
+    pub struct DataResponseV2 {
+        pub data_langid: LanguageIdentifier,
+    }
+
+    pub trait DataProviderV2<'d> {
+        fn load_v2(
```

Comment:
Should we instead have

```rust
fn load_v2<R: DataReceiver<'d, 'static>>(&self, req, receiver: &mut R);
fn load_v2_dyn(&self, req, receiver: &mut DataReceiver<..>);
```

with the latter written as a default implementation? This way for performance you can avoid the virtual dispatch.

Reply:
Potential issues with this approach:

- Pretty sure that generics of any kind aren't allowed on functions of traits that are intended to be used as trait objects.
- I'm concerned about code size explosion, since we have many possible concrete data struct types.
- Is dynamic dispatch so bad that we should design around it?  (I haven't measured it)

Reply:
Generic functions are allowed, provided you `where Self: Sized` them.

It will only be a code size explosion if used, and I think the cost of dynamic dispatch is not _high_, but it's Rusty to give people the option. And I feel like in our code we will see some cases where we prefer that.

Reply:
I will follow up here: https://github.com/unicode-org/icu4x/issues/422